### DISCUSSION TITLE
Best way to use services into twig ?

### DISCUSSION THREAD

**Discussion by a-vallet** [symfony/symfony#50523] [Q&A] [answered]
Hello everyone,

I've a question for you, we have into our professional project several way to use services into Twig, some are into Twig config file with global, others with Twig Extensions.

Each way to do it are described into the doc : 
https://symfony.com/doc/current/templates.html#templating-global-variables
https://symfony.com/doc/current/templates.html#writing-a-twig-extension

So, I was asking myself, is there a best practice to do it ?

In terms of performances put it globally and load everywhere with globals, it is bad ? Or there is lazy loading or something ?

And for extensions should we put all the logic into it and return only twig templates ? To avoid logic into twig and keep front only to display data ?

Have a nice day,

**Comment by ghost** [comment]
> In terms of performances put it globally and load everywhere with globals, it is bad ? Or there is lazy loading or something ?

Docs: `The main drawback is that these services are not loaded lazily. In other words, as soon as Twig is loaded, your service is instantiated, even if you never use that global variable.`

> And for extensions should we put all the logic into it and return only twig templates ? To avoid logic into twig and keep front only to display data ?

I don't fully understand this line. 
Twig Extensions can be written as lazy: https://symfony.com/doc/current/templates.html#creating-lazy-loaded-twig-extensions if this is what you're asking.

> So, I was asking myself, is there a best practice to do it ?

AFAIC I try not to use dynamic twig globals (i.e. globals using services) and extensions but pass all required data when rendering a template (the logic required to prepare template data is executed in dedicated factories, view model etc.). From my experience, in the end twig extensions are just overhead that needs to be maintained.

**Comment by George-Gabro** [comment, marked as answer]
You can still use extentions but don't put any logical functions in them (services etc), instead make a new file that lazy load your logical functions that extends from "RuntimeExtensionInterface". Don't forget to register your "RuntimeExtension" in services.yaml as "twig.runtime"

This will lazy load the functions that you call from twig files instead of letting symfony load every single extension before loading any page

**Reply by George-Gabro** [reply]
```php
use Symfony\Component\DependencyInjection\RuntimeExtensionInterface;

class MyCustomExtensionRuntime implements RuntimeExtensionInterface
{
    public function myBigFunction($container)
    {
    }
}

```

```
services:
    App\Runtime\MyCustomExtensionRuntime:
        tags:
            - { name: 'container.runtime' }

```
and from your function make a new twig function that calls it from MyCustomExtensionRuntime
example: `[new TwigFunction('my_custom_function', [MyCustomExtensionRuntime:class, 'myCustomFunction']),];`

**Reply by a-vallet** [reply]
Perfect i was already using extension with Runtime ;)

Thanks for the answer and all the details

**Reply by George-Gabro** [reply]
You did it the right way ;)
Glad I helped :)
